crates/iota-types/src/gas_model/gas_v2.rs

```diff
@@ -0,0 +1,522 @@
+                return Err(UserInputError::GasBudgetTooLow {
+                    gas_budget,
+                    min_budget: self.cost_table.min_transaction_cost,
+                });
+            }
+
+            // 3. Gas balance (all gas coins together) is bigger or equal to budget
+            let mut gas_balance = 0u128;
+            for gas_obj in gas_objs {
+                // expect is safe because we already checked that all gas objects have an
+                // address owner
+                gas_balance +=
+                    gas::get_gas_balance(gas_obj.as_object().expect("object must be owned"))?
+                        as u128;
+            }
+            if gas_balance < gas_budget as u128 {
+                Err(UserInputError::GasBalanceTooLow {
+                    gas_balance,
+                    needed_gas_amount: gas_budget as u128,
+                })
+            } else {
+                Ok(())
+            }
+        }
+
+        fn storage_cost(&self) -> u64 {
+            self.storage_gas_units()
+        }
+
+        pub fn per_object_storage(&self) -> &Vec<(ObjectID, PerObjectStorage)> {
+            &self.per_object_storage
+        }
+    }
+
+    impl IotaGasStatusAPI for IotaGasStatus {
+        fn is_unmetered(&self) -> bool {
+            !self.charge
+        }
+
+        fn move_gas_status(&self) -> &GasStatus {
+            &self.gas_status
+        }
+
+        fn move_gas_status_mut(&mut self) -> &mut GasStatus {
+            &mut self.gas_status
+        }
+
+        fn bucketize_computation(&mut self) -> Result<(), ExecutionError> {
+            let mut computation_units = self.gas_status.gas_used_pre_gas_price();
+            if let Some(gas_rounding) = self.gas_rounding_step {
+                if computation_units == 0 || computation_units % gas_rounding > 0 {
+                    computation_units = ((computation_units / gas_rounding) + 1) * gas_rounding;
+                }
+            } else {
+                computation_units = get_bucket_cost(&self.cost_table.computation_bucket, computation_units);
+            };
+            let computation_cost = computation_units * self.gas_price;
+            if self.gas_budget <= computation_cost {
+                self.computation_cost = self.gas_budget;
+                Err(ExecutionErrorKind::InsufficientGas.into())
+            } else {
+                self.computation_cost = computation_cost;
+                Ok(())
+            }
+        }
+
+        /// Returns the final (computation cost, storage cost, storage rebate)
+        /// of the gas meter. We use initial budget, combined with
+        /// remaining gas and storage cost to derive computation cost.
+        fn summary(&self) -> GasCostSummary {
+            // compute computation cost burned and storage rebate, both rebate and non refundable fee
+            let computation_cost_burned = self.computation_cost * self.reference_gas_price / self.gas_price;
```

Comment:
If this is the only change, then we don't need gas_v2. Maybe an `if fixed_base_fee` in gas_v1 would be sufficient

Reply:
@cyberphysic4l pls, check also this!

Reply:
I have reverted the addition of gas_v2, and in fact I don't think we even need an `if fixed_base_fee` because the computation_cost_burned field is actually unused in protocol v1, so it doesn't matter what we set for this value. So we avoid an additional if statement here.